File: leetcode/topk.go
// 利用快排算法的partition函数实现O(n)时间复杂度求数组中第k大的数，golang实现。

package main

import "fmt"

func partition(a []int, left, right int) int {
	low, up := left+1, right
	x := a[left]
	for low <= up {
		for a[low] <= x && low < up {
			low++
		}
		for a[up] > x {
			up--
		}
		if low >= up {
			break
		}
		a[low], a[up] = a[up], a[low]
	}
	a[left], a[up] = a[up], a[left]
	return up
}
func getTopK(a []int, k int) int {
	if k < 1 || k > len(a) {
		return -1
	}
	k = len(a) - k
	low, up, index := 0, len(a)-1, -1
	for index != k {
		index = partition(a, low, up)
		if index < k {
			low = index + 1
		} else {
			up = index - 1
		}
	}
	return a[index]
}
func main() {
	a := []int{7,6,5,4,3,2,1}
	k := 5
	fmt.Println("top k:", k, "value:", getTopK(a, k))
}

File: leetcode/TopK.java
// 利用快排算法的partition函数实现O(n)时间复杂度求数组中第k小的数，java实现。

class TopK {
    public static void main(String[] args) {
        int[] a = {2, 1};
        System.out.println(getTopK(a, 1));
    }

    public static int getTopK(int[] a, int k) {
        if (k < 1 || k > a.length) {
            return -1;
        }
        k = a.length - k;
        int i = 0, j = a.length - 1, index = -1;
        while (index != k) {
            index = partition(a, i, j);
            if (index > k) {
                j = index - 1;
            } else {
                i = index + 1;
            }
        }
        return a[index];
    }

    public static int partition(int[] a, int i, int j) {
        int b = i;
        int e = j + 1;
        int x = a[i];
        while (i <= j) {
            while (b < j && a[++b] <= x)
                ;
            while (a[--e] > x)
                ;
            if (b >= e) {
                break;
            }
            int tmp = a[b];
            a[b] = a[e];
            a[e] = tmp;
        }
        a[i] = a[e];
        a[e] = x;
        return e;
    }
}